### ISSUE TITLE
net: documentation for ListenIP is incorrect

### ISSUE THREAD

**Issue by perillo** [golang/go#36436]
The documentation of `net.ListenIP` says:
```
The network must be an IP network name; see func Dial for details.
```

However using one of the documented IP network names (like "ip" or "ip4") causes the function to return an error on Linux; e.g. `unknown network ip`.

According to #7439:
```
The first argument of ListenIP should be "network"+":"+"protocol on top of the IPv4 or
IPv6".
````
I confirmed that this works on Linux, but the function is clearly not documented correctly. From #7439 I assume that the format of `network` is platform dependant.

**Comment by toothrot** [comment]
@perillo Thanks for looking into this. 

The following works for me on Linux:

```
package main

import (
	"net"
)

func main() {
	conn, err := net.ListenIP("ip:icmp", &net.IPAddr{IP: net.ParseIP("127.0.0.1")})
	if err != nil {
		panic(err)
	}
	defer conn.Close()
}
```

I think what may be confusing here is the error message when the protocol is omitted, which is, as you said, `unknown network ip`. I believe this is coming from the requirement of specifying a valid protocol as read from `/etc/protocols`, in this function: 
https://github.com/golang/go/blob/77c13021dd401945711ff40f16a5040a075fcef9/src/net/dial.go#L174-L177

The documentation points to `func Dial`, which states:
>For IP networks, the network must be "ip", "ip4" or "ip6" followed by a colon and a literal protocol number or a protocol name, and the address has the form "host". The host must be a literal IP address or a literal IPv6 address with zone. It depends on each operating system how the operating system behaves with a non-well known protocol number such as "0" or "255".

Does that make sense? How could we improve this? Out of curiosity, what are you using `ListenIP` for instead of `Listen`?

/cc @bradfitz @ianlancetaylor

**Comment by perillo** [comment]
@toothrot, I was using the ListenIP function because I totally missed the second part of the Dial function documentation about IP networks.

The documentation is correct and probably is just the error message that should be more explicit.
I was probably confused by #7439, since according to @jikai507, on Windows the protocol number or protocol name is not required.

Thanks.

**Comment by toothrot** [comment]
@perillo Thanks!

Just to restate, it sounds like the error message for `net.ListenIP("ip", ...) // missing :protocol` is what was causing the confusion. Let me know if I misunderstood something, or should update the title further.

**Comment by perillo** [comment]
@toothrot, yes.  The current message is `unknown network ip`, however the problem is the missing protocol number/name.

Another issue is to check if the `network` parameter is *portable* for IP networks, due to #7439.

**Comment by toothrot** [comment]
The answer, as far as I can tell, is that the `network` parameter is platform dependent on the raw sockets API (IPConn). 

https://github.com/golang/go/blob/9341fe073e6f7742c9d61982084874560dac2014/src/net/iprawsock.go#L12-L28

On linux or similar systems, I believe it uses `/etc/protocols`. On other systems, there are various implementations, some of which end up using this map:
https://github.com/golang/go/blob/9341fe073e6f7742c9d61982084874560dac2014/src/net/lookup.go#L29-L55

I'm not sure if I can definitively answer a question about the portability of the `network` parameter in `iprawsock.go` and friends, but others may be able to.

**Comment by perillo** [comment]
@toothrot, note that services is used for looking at the default TCP/UDP port for common services.

Looking at the `network` parameter for IP network in the net package source code, it is platform dependant:
https://github.com/golang/go/blob/9341fe073e6f7742c9d61982084874560dac2014/src/net/dial.go#L173

In iprawsock.go, parseNetwork is called with needProto = false, and in iprawsock_posix.go it is called with needProto = true.
The systems considered Posix here are:
aix darwin dragonfly freebsd js,wasm linux netbsd openbsd solaris windows

I assume that in #7439, windows was not included in the build tags, so needProto was set to false, causing the call to the Windows API to fail with EPROTONOSUPPORT.

So the documentation is probably correct, since all know systems need the protocol number/name.

**Comment by perillo** [comment]
I tested on Windows 10 amd64 with the latest version of go.

 1. Without the colon (`"ip4"`) the error is `unknown network ip4`.

 2. With the colon but empty protocol (`"ip4:"`) the error is `lookup: getprotobyname: The requested name is valid, but no data of the requested type was found`.  By the way, on Linux the error is `unknown IP protocol specified`.

3. With the tcp protocol (`"ip4:tcp"`), the error is `bind: An invalid argument was supplied`.

**Comment by bradfitz** [comment]
What about `"ip4:6"`?

I wonder if we're trying to parse `/etc/protocols` on Windows and it doesn't exist?

**Comment by perillo** [comment]
@bradfitz: it is the same error `bind: An invalid argument was supplied`.

Other protocol 0, 1 and 17 works.  Also `"ip4:udp"` works.

The cause seems to be a Windows restriction:
https://docs.microsoft.com/en-us/windows/win32/winsock/tcp-ip-raw-sockets-2
`A call to the bind function with a raw socket for the IPPROTO_TCP protocol is not allowed.`

**Comment by perillo** [comment]
This issue can probably be closed.  Sorry for the noise.

**Comment by vikash-s1** [comment]
Another Issue I am facing. Client (DialIP with ip4:tcp)  -> closed Server Port (retrun FIN-ACK)
Read() is not capturing FIN-ACK packet.

conn, err := net.DialIP("ip4:tcp", &localaddr, &remoteaddr)
conn.Read(packet_data)
